File: src/scaffolds/backends/flask/services/file_service.py
"""文件服务"""
import os
import uuid
from datetime import datetime
from config import Config
from services.user_service import BizError


class FileService:

    @staticmethod
    def upload(file):
        """上传单个文件"""
        _validate(file)
        url = _save(file)
        return {'url': url, 'fileName': file.filename}

    @staticmethod
    def upload_batch(files):
        """批量上传文件"""
        if not files:
            raise BizError(400, '请选择要上传的文件')

        success, fail = [], []
        for f in files:
            try:
                _validate(f)
                url = _save(f)
                success.append({'fileName': f.filename, 'url': url})
            except Exception as e:
                fail.append(f'{f.filename}: {str(e)}')

        return {
            'success': success, 'fail': fail,
            'total': len(files), 'successCount': len(success), 'failCount': len(fail)
        }

    @staticmethod
    def delete(file_name):
        """删除文件"""
        if not file_name:
            raise BizError(400, '文件名不能为空')
        relative_path = file_name.replace('/uploads/', '')
        base_path = os.path.realpath(Config.UPLOAD_DIR)
        absolute_path = os.path.realpath(os.path.join(base_path, relative_path))
        if not absolute_path.startswith(base_path + os.sep) and absolute_path != base_path:
            raise BizError(400, '非法的文件路径')
        if os.path.exists(absolute_path):
            os.remove(absolute_path)


def _validate(file):
    if not file or not file.filename:
        raise BizError(400, '文件不能为空')
    ext = file.filename.rsplit('.', 1)[-1].lower() if '.' in file.filename else ''
    if ext not in Config.ALLOWED_EXTENSIONS:
        raise BizError(400, f'不支持的文件类型: .{ext}')


def _save(file) -> str:
    ext = file.filename.rsplit('.', 1)[-1].lower()
    new_name = f'{uuid.uuid4().hex}.{ext}'
    date_dir = datetime.now().strftime('%Y-%m-%d')
    dir_path = os.path.join(Config.UPLOAD_DIR, date_dir)
    os.makedirs(dir_path, exist_ok=True)

    file_path = os.path.join(dir_path, new_name)
    file.save(file_path)
    return f'/uploads/{date_dir}/{new_name}'

File: src/scaffolds/backends/express/src/services/fileService.js
/** 文件服务 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { uploadConfig } from '../config/upload.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const UPLOAD_BASE = path.join(__dirname, '../../../../uploads');

const ALLOWED_EXTENSIONS = new Set([
  '.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp',
  '.pdf', '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx',
  '.txt', '.zip', '.rar', '.7z'
]);

// 最大文件大小统一取 uploadConfig，避免与路由层限制不一致
const MAX_SIZE = uploadConfig.maxSize;

export const fileService = {
  // 上传单个文件
  async upload(file) {
    validate(file);
    const url = await save(file);
    return { url, fileName: file.originalname };
  },

  // 批量上传文件
  async uploadBatch(files) {
    if (!files || files.length === 0) throw { code: 400, message: '请选择要上传的文件' };

    const success = [];
    const fail = [];

    for (const file of files) {
      try {
        validate(file);
        const url = await save(file);
        success.push({ fileName: file.originalname, url });
      } catch (e) {
        fail.push(`${file.originalname}: ${e.message}`);
      }
    }

    return { success, fail, total: files.length, successCount: success.length, failCount: fail.length };
  },

  // 删除文件
  delete(fileName) {
    if (!fileName) throw { code: 400, message: '文件名不能为空' };
    const relativePath = fileName.replace('/uploads/', '');
    const absolutePath = path.resolve(UPLOAD_BASE, relativePath);
    if (!absolutePath.startsWith(path.resolve(UPLOAD_BASE))) {
      throw { code: 400, message: '非法的文件路径' };
    }
    if (fs.existsSync(absolutePath)) fs.unlinkSync(absolutePath);
  }
};

function validate(file) {
  if (!file || file.size === 0) throw { code: 400, message: '文件不能为空' };
  if (file.size > MAX_SIZE) throw { code: 400, message: `文件大小不能超过${MAX_SIZE / 1024 / 1024}MB` };
  const ext = path.extname(file.originalname).toLowerCase();
  if (!ALLOWED_EXTENSIONS.has(ext)) throw { code: 400, message: `不支持的文件类型: ${ext}` };
}

async function save(file) {
  const ext = path.extname(file.originalname);
  const newName = `${Date.now()}-${Math.round(Math.random() * 1e9)}${ext}`;
  const dateDir = new Date().toISOString().split('T')[0];
  const dir = path.join(UPLOAD_BASE, dateDir);

  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

  if (file.buffer) {
    fs.writeFileSync(path.join(dir, newName), file.buffer);
  } else if (file.path) {
    fs.renameSync(file.path, path.join(dir, newName));
  }

  return `/uploads/${dateDir}/${newName}`;
}
